sample/weighted_test.go

```diff
@@ -0,0 +1,236 @@
+// Copyright ©2015 The gonum Authors. All rights reserved.
+// Use of this code is governed by a BSD-style
+// license that can be found in the LICENSE file
+
+package sample
+
+import (
+	"flag"
+	"math/rand"
+	"reflect"
+	"testing"
+	"time"
+
+	"github.com/gonum/floats"
+)
+
+var prob = flag.Bool("prob", false, "enables probabilistic testing of the random weighted sampler")
+
+const sigChi2 = 16.92 // p = 0.05 df = 9
+
+var (
+	newExp = func() []float64 {
+		return []float64{1 << 0, 1 << 1, 1 << 2, 1 << 3, 1 << 4, 1 << 5, 1 << 6, 1 << 7, 1 << 8, 1 << 9}
+	}
+	exp = newExp()
+
+	obt = []float64{973, 1937, 3898, 7897, 15769, 31284, 62176, 125408, 250295, 500363}
+)
+
+func newTestWeighted() Weighted {
+	weights := make([]float64, len(obt))
+	for i := range weights {
+		weights[i] = float64(int(1) << uint(i))
+	}
+	return NewWeighted(weights, nil)
+}
+
+func TestWeightedUnseeded(t *testing.T) {
+	rand.Seed(0)
+
+	want := Weighted{
+		weights: []float64{1 << 0, 1 << 1, 1 << 2, 1 << 3, 1 << 4, 1 << 5, 1 << 6, 1 << 7, 1 << 8, 1 << 9},
+		heap: []float64{
+			exp[0] + exp[1] + exp[3] + exp[4] + exp[7] + exp[8] + exp[9] + exp[2] + exp[5] + exp[6],
+			exp[1] + exp[3] + exp[4] + exp[7] + exp[8] + exp[9],
+			exp[2] + exp[5] + exp[6],
+			exp[3] + exp[7] + exp[8],
+			exp[4] + exp[9],
+			exp[5],
+			exp[6],
+			exp[7],
+			exp[8],
+			exp[9],
+		},
+	}
+
+	ts := newTestWeighted()
+	if !reflect.DeepEqual(ts, want) {
+		t.Fatalf("unexpected new Weighted value:\ngot: %#v\nwant:%#v", ts, want)
+	}
+
```

Comment:
Could we do a test of taking all of the items out? Get a new weighted, take each item out, make sure they're all there, and make sure that taking one more item out fails.